Before you turn this homework set in, make sure everything runs as expected. When you are done modifying this file,  **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name below. In the code cells, make sure to replace the **raise NotImplementedError()** statement with your actual code.

In [1]:
NAME = "Liyuan Sun"

---

## Problem 1

The file "Temp.txt" which you can download from the homework tab on Canvas, contains the monthly high-temperatures (in degrees Celsius) at Heathrow airport for the years 1948 through 2016. Save the file on your computer (in a folder that you can find again). Run the code below, changing the path to the appropriate path on **your** computer. It's ok to use forward slashes in the path name, regardless of whether you're working on Windows or MacOS. 

In [6]:
temperatures = []

from pathlib import Path

data_folder = Path("/Users/liyuansun/Downloads/167ps/Homework_2") 
# make sure to change this path to the actual path where YOU have saved the data file. 

file_to_open = data_folder / "Temp.txt"

f = open(file_to_open)

for row in f:
    temperatures.append(float(row))

You have now created an object called temperatures. 

Answer the following questions by inserting your code into the code box (replacing the "# your code here" prompt) and answering the question in the Markdown box. 

a) What kind of object type is "temperatures"?

In [20]:
type(temperatures)

list

YOUR ANSWER HERE

b) Describe what the "float()" and "append()" commands do, respectively, that have been applied above. Try to leave out the "float()" command. What happens then? 

- float() is a function, change the type of string into float. 

- append() is a method, add each float(row) into the list temperatures. 

- Then, if we leave out the "float()" commend, there's an error. 

c) How many temperatures have been recorded in the file?

In [46]:
len(temperatures)

828

YOUR ANSWER HERE

d) What are the highest and lowest temperatures recorded?

In [47]:
temperatures.sort()
print(temperatures[0])
print(temperatures[-1])

0.8
28.2


The highest temperature is 28.2 (in degrees Celsius).

The lowest temperature is 0.8 (in degrees Celsius).


YOUR ANSWER HERE

e) Find the mean temperature (using only the commands you learned from chapter 5, not using additional modules or libraries). Use Python to round your result to two after-decimal digits. 

In [48]:
round(sum(temperatures)/len(temperatures), 2)

14.85

YOUR ANSWER HERE

e) Find the median temperature (using only the commands you learned from chapter 5, not using additional modules or libraries).

In [50]:
L = int(len(temperatures)/2 - 1)
median = (temperatures[L] + temperatures[L + 1])/2
median

14.7

YOUR ANSWER HERE

## Problem 2 

Much of data science these days is done on text data. The text to be analyzed may be literature or postings on social media. The tasks to be performed could be sentiment analysis (trying to guess whether the writing is "pro" or "con" on a certain issue) to automatically detecting postings by Russian bots. However, before you can perform any kind of machine learning procedure on a text, you have to first learn how to "clean" the text. That usually means removing all non-text elements and sometimes also removing filler words such as "the", "it", "and" etc. After that. what is usually analyzed is the frequency with which certain words appear and most often actually not the *order* in which they appear. 

The text below are the first three paragraphs of Franz Kafka's book "Metamorphosis". Our goal is to clean this text passage. 

In [46]:
text = """One morning, when Gregor Samsa woke from troubled dreams, he found
himself transformed in his bed into a horrible vermin.  He lay on
his armour-like back, and if he lifted his head a little he could
see his brown belly, slightly domed and divided by arches into stiff
sections.  The bedding was hardly able to cover it and seemed ready
to slide off any moment.  His many legs, pitifully thin compared
with the size of the rest of him, waved about helplessly as he
looked.

"What's happened to me?" he thought.  It wasn't a dream.  His room,
a proper human room although a little too small, lay peacefully
between its four familiar walls.  A collection of textile samples
lay spread out on the table - Samsa was a travelling salesman - and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice, gilded frame.  It showed
a lady fitted out with a fur hat and fur boa who sat upright,
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer.

Gregor then turned to look out the window at the dull weather.
Drops of rain could be heard hitting the pane, which made him feel
quite sad.  "How about if I sleep a little bit longer and forget all
this nonsense", he thought, but that was something he was unable to
do because he was used to sleeping on his right, and in his present
state couldn't get into that position.  However hard he threw
himself onto his right, he always rolled back to where he was.  He
must have tried it a hundred times, shut his eyes so that he
wouldn't have to look at the floundering legs, and only stopped when
he began to feel a mild, dull pain there that he had never felt
before."""

a) Look at the text. What non-alphabetic character elements do you see in the text?

The non-alphabetic character elements in the text are ,."'?-

b) Using only the commands covered in class so far, write code to remove all non-alphabetic characters (such as punctuation).

In [47]:
text = text.replace(",","")
text = text.replace(".","")
text = text.replace("\"","")
text = text.replace("'","")
text = text.replace("?","")
text = text.replace("-","")
print(text)

One morning when Gregor Samsa woke from troubled dreams he found
himself transformed in his bed into a horrible vermin  He lay on
his armourlike back and if he lifted his head a little he could
see his brown belly slightly domed and divided by arches into stiff
sections  The bedding was hardly able to cover it and seemed ready
to slide off any moment  His many legs pitifully thin compared
with the size of the rest of him waved about helplessly as he
looked

Whats happened to me he thought  It wasnt a dream  His room
a proper human room although a little too small lay peacefully
between its four familiar walls  A collection of textile samples
lay spread out on the table  Samsa was a travelling salesman  and
above it there hung a picture that he had recently cut out of an
illustrated magazine and housed in a nice gilded frame  It showed
a lady fitted out with a fur hat and fur boa who sat upright
raising a heavy fur muff that covered the whole of her lower arm
towards the viewer

Gregor 

c) Write code to make everything lower case and convert the text into a list of single words. 

In [49]:
text = text.lower()
textList = text.split()
print(textList)

['one', 'morning', 'when', 'gregor', 'samsa', 'woke', 'from', 'troubled', 'dreams', 'he', 'found', 'himself', 'transformed', 'in', 'his', 'bed', 'into', 'a', 'horrible', 'vermin', 'he', 'lay', 'on', 'his', 'armourlike', 'back', 'and', 'if', 'he', 'lifted', 'his', 'head', 'a', 'little', 'he', 'could', 'see', 'his', 'brown', 'belly', 'slightly', 'domed', 'and', 'divided', 'by', 'arches', 'into', 'stiff', 'sections', 'the', 'bedding', 'was', 'hardly', 'able', 'to', 'cover', 'it', 'and', 'seemed', 'ready', 'to', 'slide', 'off', 'any', 'moment', 'his', 'many', 'legs', 'pitifully', 'thin', 'compared', 'with', 'the', 'size', 'of', 'the', 'rest', 'of', 'him', 'waved', 'about', 'helplessly', 'as', 'he', 'looked', 'whats', 'happened', 'to', 'me', 'he', 'thought', 'it', 'wasnt', 'a', 'dream', 'his', 'room', 'a', 'proper', 'human', 'room', 'although', 'a', 'little', 'too', 'small', 'lay', 'peacefully', 'between', 'its', 'four', 'familiar', 'walls', 'a', 'collection', 'of', 'textile', 'samples', 'l

d) How many words are in the list? Count, how often the word "he" appears in the three paragraphs. 

In [50]:
print(len(textList))
print(textList.count("he"))

313
17


YOUR ANSWER HERE

e) How many distinct (different) words appear in the first three paragragraphs?

In [52]:
textSet = set(textList)
len(textSet)

194

YOUR ANSWER HERE